iceoryx_posh/test/moduletests/test_popo_server_port_user.cpp

```diff
@@ -0,0 +1,732 @@
+        EXPECT_THAT(this->getNumberOfUsedChunks(), Eq(NUMBER_OF_REQUEST_CHUNKS));
+    });
+}
+
+TEST_F(ServerPort_test, FreeResponseWithInvalidChunkCallsTheErrorHandler)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "fb09c788-0e1b-41c6-877a-1b13a37829d4");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    iox::cxx::optional<iox::Error> detectedError;
+    auto errorHandlerGuard = iox::ErrorHandler::setTemporaryErrorHandler(
+        [&](const iox::Error error, const std::function<void()>, const iox::ErrorLevel errorLevel) {
+            EXPECT_THAT(error, Eq(iox::Error::kPOPO__CHUNK_SENDER_INVALID_CHUNK_TO_FREE_FROM_USER));
+            EXPECT_THAT(errorLevel, Eq(iox::ErrorLevel::SEVERE));
+            detectedError.emplace(error);
+        });
+
+    allocateResponseWithRequestHeaderAndThen(sut, [&](const auto, auto res) {
+        sut.portUser.freeResponse(res);
+        // since the response is already freed, it should not be in the UsedChunkList anymore and the error handler
+        // should be called
+        sut.portUser.freeResponse(res);
+    });
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+TEST_F(ServerPort_test, FreeResponseWithWithNullptrResponseHeaderCallsTheErrorHandler)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ecb40c4d-7b95-4780-9b51-ac1708830453");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    iox::cxx::optional<iox::Error> detectedError;
+    auto errorHandlerGuard = iox::ErrorHandler::setTemporaryErrorHandler(
+        [&](const iox::Error error, const std::function<void()>, const iox::ErrorLevel errorLevel) {
+            EXPECT_THAT(error, Eq(iox::Error::kEXPECTS_ENSURES_FAILED));
+            EXPECT_THAT(errorLevel, Eq(iox::ErrorLevel::FATAL));
+            detectedError.emplace(error);
+        });
+
+    sut.portUser.freeResponse(nullptr);
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+// END freeResponse tests
+
+// BEGIN sendResponse tests
+
+TEST_F(ServerPort_test, SendResponseWithWithNullptrResponseHeaderCallsTheErrorHandler)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8d0b0a7f-d1ab-4249-b885-cbc6429eab83");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    iox::cxx::optional<iox::Error> detectedError;
+    auto errorHandlerGuard = iox::ErrorHandler::setTemporaryErrorHandler(
+        [&](const iox::Error error, const std::function<void()>, const iox::ErrorLevel errorLevel) {
+            EXPECT_THAT(error, Eq(iox::Error::kEXPECTS_ENSURES_FAILED));
+            EXPECT_THAT(errorLevel, Eq(iox::ErrorLevel::FATAL));
+            detectedError.emplace(error);
+        });
+
+    sut.portUser.sendResponse(nullptr);
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+TEST_F(ServerPort_test, SendResponseWithWithoutOfferReleasesTheChunkToTheMempool)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "dc4e31b1-18bf-4c42-9084-dd7abd52609b");
+    auto& sut = serverPortWithoutOfferOnCreate;
+
+    allocateResponseWithRequestHeaderAndThen(sut, [&](const auto, auto res) {
```

Comment:
Here I would prefer to inform the user that the send has not succeeded. Could we return a `bool` in `sendResponse` which returns `false` when the port is not in the offer state.

Reply:
This again boils down to the question if we want familiarity with the pub-sub API?

Reply:
No, I see a good API as a process and we converge with every feature and every PR to a more fitting API which covers better and better our use case.
Therefore, I would say we start with the pub-sub API here but develop it further and when this plays out well we create an issue to refine the pub sub API.

With this we improve the API step by step since we cannot think of everything right from the beginning.

Reply:
To summarize for `getRequest` I follow your argumentation and you can leave it has you mentioned it earlier but here I would really like to inform the user. What if a server allocates some resource and thinks, because of the lack of error, that this resource was send to the client.

Reply:
not yet done, but the comment below has the same request and the discussion happens there